src/Cli/dotnet/commands/dotnet-workload/restore/WorkloadRestoreCommand.cs

```diff
@@ -31,15 +33,46 @@ public WorkloadRestoreCommand(
 
         public override int Execute()
         {
-            var allProjects = DiscoverAllProjects(Directory.GetCurrentDirectory(), _slnOrProjectArgument).Distinct();
-            List<WorkloadId> allWorkloadId = RunTargetToGetWorkloadIds(allProjects);
-            Reporter.WriteLine(string.Format(LocalizableStrings.InstallingWorkloads, string.Join(" ", allWorkloadId)));
+            var globalJsonPath = SdkDirectoryWorkloadManifestProvider.GetGlobalJsonPath(Environment.CurrentDirectory);
+            var workloadSetVersionFromGlobalJson = SdkDirectoryWorkloadManifestProvider.GlobalJsonReader.GetWorkloadVersionFromGlobalJson(globalJsonPath);
+
+            var workloadResolverFactory = new WorkloadResolverFactory();
+            var creationResult = workloadResolverFactory.Create();
+            var workloadInstaller = WorkloadInstallerFactory.GetWorkloadInstaller(NullReporter.Instance, new SdkFeatureBand(creationResult.SdkVersion),
+                                        creationResult.WorkloadResolver, Verbosity, creationResult.UserProfileDir, VerifySignatures, PackageDownloader,
+                                        creationResult.DotnetPath, TempDirectoryPath, null, RestoreActionConfiguration, elevationRequired: true);
+            var recorder = new WorkloadHistoryRecorder(
+                               creationResult.WorkloadResolver,
+                               workloadInstaller,
+                               () => workloadResolverFactory.CreateForWorkloadSet(
+                                   creationResult.DotnetPath,
+                                   creationResult.SdkVersion.ToString(),
+                                   creationResult.UserProfileDir,
+                                   null));
+            recorder.HistoryRecord.CommandName = "restore";
+
+            recorder.Run(() =>
+            {
+                // If there's a workload set version specified in the global.json file, we need to make sure the workload set is installed before we try to discover workload ids.
+                if (!string.IsNullOrWhiteSpace(workloadSetVersionFromGlobalJson))
+                {
+                    if (creationResult.WorkloadResolver.GetWorkloadManifestProvider() is SdkDirectoryWorkloadManifestProvider provider && provider.WouldThrowException())
```

Comment:
Rather than casting to the implementation class, if we need a new method we should add it to the interface.